diff --git a/board/compulab/plat/imx8mm/board/board.c b/board/compulab/plat/imx8mm/board/board.c
index e9ea89886f..b08252976d 100644
--- a/board/compulab/plat/imx8mm/board/board.c
+++ b/board/compulab/plat/imx8mm/board/board.c
@@ -417,7 +417,12 @@ int board_mmc_get_env_dev(int devno)
 		printf("User Environment dev# is (%lu)\n", user_env_devno);
 		return (int)user_env_devno;
 	}
-	return devno;
+	/* mach-imx/imx8/cpu.c passes devno as follows
+	     0 = SC_R_SDHC_0 (not available on our boards)
+	     1 = SC_R_SDHC_1 (mmc0 for our boards)
+	     2 = SC_R_SDHC_2 (mmc1 for our boards)
+	   thus we need to adjust devno */
+	return devno - 1;
 }
 
 static int _mmc_get_env_part(struct mmc *mmc)
diff --git a/board/compulab/plat/imx8mm/board/board_addon.c b/board/compulab/plat/imx8mm/board/board_addon.c
index 088ae05479..5fed46af21 100644
--- a/board/compulab/plat/imx8mm/board/board_addon.c
+++ b/board/compulab/plat/imx8mm/board/board_addon.c
@@ -6,9 +6,12 @@
 
 #include <common.h>
 #include <led.h>
+#include <asm-generic/gpio.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#ifdef CONFIG_LED
+
 static int led_test_on_off(int duration_ms)
 {
 	struct udevice *dev;
@@ -47,7 +50,7 @@ static int test_system_leds(void)
 	return 0;
 }
 
-static int init_system_status_led(void)
+static int init_system_status_led(int state)
 {
 	struct udevice *dev;
 	char *led_name;
@@ -59,7 +62,7 @@ static int init_system_status_led(void)
 		if (ret)
 			return ret;
 
-		ret = led_set_state(dev, LEDST_ON);
+		ret = led_set_state(dev, state);
 		if (ret)
 			return ret;
 	}
@@ -67,6 +70,41 @@ static int init_system_status_led(void)
 	return 0;
 }
 
+#else
+
+static int test_system_leds(void)
+{
+	return 0;
+}
+
+static int init_system_status_led(int state)
+{
+	return 0;
+}
+
+#endif
+
+static void init_fastboot_variables()
+{
+	const char* model = NULL;
+	char cpu_name[64];
+	u32 cpurev = get_cpu_rev();
+
+	model = fdt_getprop(gd->fdt_blob, 0, "model", NULL);
+	snprintf(cpu_name, sizeof(cpu_name), "Freescale i.MX%s rev%d.%d",
+		get_imx_type((cpurev & 0xFF000) >> 12), (cpurev & 0x000F0) >> 4, (cpurev & 0x0000F) >> 0);
+
+	if(model)
+		env_set("board", model);
+	env_set("platform", cpu_name);
+	//env_set("fastboot.serialno", );
+#if defined(CONFIG_FIT_SIGNATURE) && !defined(CONFIG_LEGACY_IMAGE_FORMAT)
+	env_set("fastboot.secure", "yes");
+#else
+	env_set("fastboot.secure", "no");
+#endif
+}
+
 static int setup_mac_address_x(int x)
 {
 	int ret;
@@ -94,11 +132,10 @@ int misc_init_r(void)
 {
 	int x, ret;
 
-	if (IS_ENABLED(CONFIG_LED)) {
-		test_system_leds();
-		led_default_state();
-		init_system_status_led();
-	}
+	test_system_leds();
+	led_default_state();
+	init_system_status_led(LEDST_ON);
+	init_fastboot_variables();
 
 	for (x=0; x<4; x++) {
 		ret = setup_mac_address_x(x);
diff --git a/cmd/bootmenu.c b/cmd/bootmenu.c
index 3dc2c854ac..1d79e773fd 100644
--- a/cmd/bootmenu.c
+++ b/cmd/bootmenu.c
@@ -206,6 +206,10 @@ static void bootmenu_loop(struct bootmenu_data *menu,
 	if (c == '\r')
 		*key = KEY_SELECT;
 
+	/* netconsole sends only line feed in boot menu, so check for 0xA */
+	if (c == 0xA)
+		*key = KEY_SELECT;
+
 	/* ^C was pressed */
 	if (c == 0x3)
 		*key = KEY_QUIT;
@@ -277,7 +277,7 @@ static void bootmenu_destroy(struct bootmenu_data *menu)
 	free(menu);
 }
 
-static struct bootmenu_data *bootmenu_create(int delay)
+static struct bootmenu_data *bootmenu_create(int delay, int console)
 {
 	unsigned short int i = 0;
 	const char *option;
@@ -347,40 +349,41 @@ static struct bootmenu_data *bootmenu_create(int delay)
 			break;
 	}
 
-	/* Add U-Boot console entry at the end */
-	if (i <= MAX_COUNT - 1) {
-		entry = malloc(sizeof(struct bootmenu_entry));
-		if (!entry)
-			goto cleanup;
-
-		entry->title = strdup("U-Boot console");
-		if (!entry->title) {
-			free(entry);
-			goto cleanup;
-		}
+	if (console != 0) {
+		/* Add U-Boot console entry at the end */
+		if (i <= MAX_COUNT - 1) {
+			entry = malloc(sizeof(struct bootmenu_entry));
+			if (!entry)
+				goto cleanup;
+
+			entry->title = strdup("U-Boot console");
+			if (!entry->title) {
+				free(entry);
+				goto cleanup;
+			}
 
-		entry->command = strdup("");
-		if (!entry->command) {
-			free(entry->title);
-			free(entry);
-			goto cleanup;
-		}
+			entry->command = strdup("");
+			if (!entry->command) {
+				free(entry->title);
+				free(entry);
+				goto cleanup;
+			}
 
-		sprintf(entry->key, "%d", i);
+			sprintf(entry->key, "%d", i);
 
-		entry->num = i;
-		entry->menu = menu;
-		entry->next = NULL;
+			entry->num = i;
+			entry->menu = menu;
+			entry->next = NULL;
 
-		if (!iter)
-			menu->first = entry;
-		else
-			iter->next = entry;
+			if (!iter)
+				menu->first = entry;
+			else
+				iter->next = entry;
 
-		iter = entry;
-		++i;
+			iter = entry;
+			++i;
+		}
 	}
-
 	menu->count = i;
 
 	if ((menu->active >= menu->count)||(menu->active < 0)) { //ensure active menuitem is inside menu
@@ -423,7 +426,7 @@ cleanup:
 	puts(ANSI_CLEAR_LINE);
 }
 
-static void bootmenu_show(int delay)
+static void bootmenu_show(int delay, int console)
 {
 	int init = 0;
 	void *choice = NULL;
@@ -450,7 +453,7 @@ static void bootmenu_show(int delay)
 		return;
 	}
 
-	bootmenu = bootmenu_create(delay);
+	bootmenu = bootmenu_create(delay, console);
 	if (!bootmenu)
 		return;
 
@@ -505,9 +508,9 @@ void menu_display_statusline(struct menu *m)
 }
 
 #ifdef CONFIG_AUTOBOOT_MENU_SHOW
-int menu_show(int bootdelay)
+int menu_show(int bootdelay, int console)
 {
-	bootmenu_show(bootdelay);
+	bootmenu_show(bootdelay, console);
 	return -1; /* -1 - abort boot and run monitor code */
 }
 #endif
@@ -516,6 +519,7 @@ int do_bootmenu(cmd_tbl_t *cmdtp, int flag, int argc, char *const argv[])
 {
 	char *delay_str = NULL;
 	int delay = 10;
+	int console = 0;
 
 #if defined(CONFIG_BOOTDELAY) && (CONFIG_BOOTDELAY >= 0)
 	delay = CONFIG_BOOTDELAY;
@@ -524,19 +528,23 @@ int do_bootmenu(cmd_tbl_t *cmdtp, int flag, int argc, char *const argv[])
 	if (argc >= 2)
 		delay_str = argv[1];
 
+	if (argc >= 3)
+		console = (strcmp(argv[2],"0") == 0) ? 0 : 1;
+
 	if (!delay_str)
 		delay_str = env_get("bootmenu_delay");
 
 	if (delay_str)
 		delay = (int)simple_strtol(delay_str, NULL, 10);
 
-	bootmenu_show(delay);
+	bootmenu_show(delay, console);
 	return 0;
 }
 
 U_BOOT_CMD(
-	bootmenu, 2, 1, do_bootmenu,
+	bootmenu, 3, 1, do_bootmenu,
 	"ANSI terminal bootmenu",
 	"[delay]\n"
+	"[0/1] add console entry\n"
 	"    - show ANSI terminal bootmenu with autoboot delay"
 );
diff --git a/drivers/fastboot/fb_mmc.c b/drivers/fastboot/fb_mmc.c
index d7cf9f4aac..ccacd23432 100644
--- a/drivers/fastboot/fb_mmc.c
+++ b/drivers/fastboot/fb_mmc.c
@@ -589,6 +589,18 @@ void fastboot_mmc_flash_write(const char *cmd, void *download_buffer,
 	}
 #endif
 
+	if (strcmp(cmd, "bootloader") == 0) {
+		info.blksz = dev_desc->blksz;
+		info.start = (33 * SZ_1K) / info.blksz;
+		info.size  = (2 * SZ_1M) / info.blksz;
+		printf("%s: updating bootloader\n", __func__);
+		write_raw_image(dev_desc, &info, cmd, download_buffer,
+				download_bytes, response);
+		printf("........ success\n");
+		fastboot_okay(NULL, response);
+		return;
+	}
+
 #ifdef CONFIG_ANDROID_BOOT_IMAGE
 	if (strncasecmp(cmd, "zimage", 6) == 0) {
 		dev_desc = fastboot_mmc_get_dev(response);
